File: python/merge_sort.py
import sort_tools
import copy

def merge(data, low, middle, high):
    data_a = []
    data_b = []
    len_a = middle - low + 1
    len_b = high - middle
    i = 0   # track data_a
    j = 0   # track data_b
    k = low # track original data set

    for idx in range(middle - low + 1):
        data_a.append(data[low + idx])

    for idx in range(high - middle):
        data_b.append(data[middle + idx + 1])

    #compare and merge
    while i < len_a and j < len_b:
        a = data[i]
        b = data[j]

        if a < b:
            data[k] = a
            i += 1
        else:
            data[k] = b
            j += 1
        k += 1

    #merge any trailing elements
    while i < len_a:
        data[k] = data_a[i]
        i += 1
        k += 1

    while j < len_b:
        data[k] = data_b[j]
        j += 1
        k += 1
    
def merge_sort(data, low, high):
    if low < high:
        middle = (high - low) // 2 + low
        merge_sort(data, low, middle)
        merge_sort(data, middle + 1, high)
        merge(data, low, middle, high)

@sort_tools.timeit('merge')
def sort(data):
    merge_sort(data, 0, len(data) - 1)

def main():
    data = sort_tools.build_data_set(5)

    unsorted = copy.deepcopy(data)
    sort(unsorted)

if __name__ == "__main__":
    main()

File: Java/MergeSort.java
package Java;

public class MergeSort extends Sort{
	
	public static void main(String[] args){
		int[] data = buildDataSet(5);
		
		timeSort(new MergeSort(), data);
	}
	
	public MergeSort(){ super("Merge"); }
	
	private int[] merge(int[] data, int low, int middle, int high){
		int i = 0; 		// track dataA's index
		int j = 0; 		// track dataB's index
		int k = low; 	// track data's index
		int lenA = middle - low + 1;
		int lenB = high - middle;
		
		int[] dataA = new int[lenA];
		int[] dataB = new int[lenB];
		
		for (int idx = 0; idx < lenA; idx++)
			dataA[idx] = data[low + idx];
		
		for (int idx = 0; idx < lenB; idx++)
			dataB[idx] = data[middle + idx + 1];
		
		//compare and merge
		while ( i < lenA && j < lenB )
		{
			if ( dataA[i] < dataB[j] )
				data[k++] = dataA[i++];
			else
				data[k++] = dataB[j++];
		}
		
		//merge any trailing elements
		while ( i < dataA.length )
			data[k++] = dataA[i++];
		
		while ( j < dataB.length )
			data[k++] = dataB[j++];
		
		return data;
	}
	
	private void mergeSort(int[] data, int low, int high){
		if ( low < high )
		{
			int middle = (high - low) / 2 + low;
			mergeSort(data, low, middle);
			mergeSort(data, middle + 1, high);
			merge(data, low, middle, high);
		}
	}
	
	@Override
	public void sort(int[] data){
		mergeSort(data, 0, data.length - 1);
	}
}
